Library/OcCryptoLib/RsaDigitalSign.c

```diff
@@ -0,0 +1,511 @@
+  // Exponent 65537
+  //
+  for (Index = 0; Index < 16; Index += 2) {
+    //
+    // Aar = Ar * Ar / R mod M 
+    //
+    MontMul (Key, Aar, Ar, Ar);
+    //
+    // Ar = Aar * Aar / R mod M 
+    //
+    MontMul (Key, Ar, Aar, Aar);
+  }
+  //
+  // Aaa = Ar * A / R mod M
+  //
+  MontMul (Key, Aaa, Ar, A);
+
+  //
+  // Make sure Aaa < Mod; Aaa is at most 1x mod too large.
+  //
+  if (GeMod (Key, Aaa)){
+    SubMod (Key, Aaa);
+  }
+
+  //
+  // Convert to bigendian byte array
+  //
+  for (Index = (INT32) Key->Size - 1; Index >= 0; --Index) {
+    Tmp = Aaa[Index];
+
+    *InOut++ = (UINT8) (Tmp >> 24);
+    *InOut++ = (UINT8) (Tmp >> 16);
+    *InOut++ = (UINT8) (Tmp >>  8);
+    *InOut++ = (UINT8) (Tmp >>  0);
+  }
+
+  //
+  // Free work buffers before return
+  //
+  if (A != NULL) {
+    FreePool (A);
+  }
+  if (Ar != NULL) {
+    FreePool (Ar);
+  }
+  if (Aar != NULL) {
+    FreePool (Aar);
+  }  
+}
+
+/**
+  This routine parses key data from RsaPublicKey Header
+
+  @ param KeyData         RSA public key data
+  @ param Length          RSA public key size
+ */
+STATIC
+RSA_PUBLIC_KEY
+*RsaParseKeyData (
+  UINT8  *KeyData,
+  UINTN  Length
+  )
+{
+  RSA_PUBLIC_KEY_HEADER  KeyHeader;
+  CONST UINT8            *N;
+  CONST UINT8            *Rr;
+  UINT32                 Index           = 0;
+  UINT32                 RsaKeyStructLen = 0;
+  RSA_PUBLIC_KEY         *RsaPublicKey   = NULL;
+  UINTN                  ExpectedLength  = 0;
+
+  //
+  // Copy KeyHeader
+  //
+  CopyMem (&KeyHeader, KeyData, sizeof (RSA_PUBLIC_KEY_HEADER));
+
+  /*
+  //
+  // Validate KeyHeader
+  //
+  if (!RsaPkValidateByteswape ((CONST RSA_PUBLIC_KEY_HEADER *) KeyData, &KeyHeader)) {
+    //
+    // Ooops. Invalid key
+    //
+    DEBUG ((DEBUG_INFO, "Invalid key.\n"));
+
+    if (RsaPublicKey != NULL) {
+      FreePool (RsaPublicKey);  
+    }
+    return NULL;
+  }
+  */
+
+  //
+  // Validate key num bits
+  //
+  if (    KeyHeader.KeyNumBits != 2048 
+       || KeyHeader.KeyNumBits != 4096
+       || KeyHeader.KeyNumBits != 8192 ) {
+    DEBUG ((DEBUG_INFO, "Unexpected key length.\n"));
+    if (RsaPublicKey != NULL) {
+      FreePool (RsaPublicKey);
+    }
+    return NULL;
+  }
+
+  //
+  // Calculate expected key length based on key num bits
+  //
+  ExpectedLength = sizeof (RSA_PUBLIC_KEY_HEADER) + 2 * KeyHeader.KeyNumBits / 8;
+
+  //
+  // Validate expected key length with passed length from function args
+  //
+  if (Length != ExpectedLength) {
+    DEBUG ((DEBUG_INFO, "Key does not match expected length\n"));
+    if (RsaPublicKey != NULL) {
+      FreePool (RsaPublicKey);
+    }
+    return NULL;    
+  }
+
+  //
+  // Extract N
+  //
+  N = KeyData + sizeof (RSA_PUBLIC_KEY_HEADER);
+  //
+  // Extract R^2
+  //
+  Rr = KeyData + sizeof (RSA_PUBLIC_KEY_HEADER) + KeyHeader.KeyNumBits / 8;
+
+  RsaKeyStructLen = sizeof (RSA_PUBLIC_KEY) + 2 * KeyHeader.KeyNumBits / 8;
+  RsaPublicKey = (RSA_PUBLIC_KEY *) (AllocateZeroPool (RsaKeyStructLen));
```

Comment:
As C is fine with implicit casts to and from VOID *, I think we should avoid the clutter of casting allocation results